### PR TITLE
Add Allies 09a

### PR DESCRIPTION
**Pull request by yuantse**
Add Allied campaign mission 09a

### CODE REVIEW

**Comment by Orb370** [comment]
Hello, I played through the mission on hard difficulty. I've attached the replay if you want to see it.

[Allies 09a -Orb.zip](https://github.com/OpenRA/OpenRA/files/5360523/Allies.09a.-Orb.zip)

It was a little easy. I would have the attack where enemies land at your base occur sooner (on hard). Also the enemy subs didn't attack at all (I'm not sure if this was the case in the original).

I think the helipad should be replaced with a barracks, since in openRA multiplayer Soviets do not have access to the helipad.  @Smittytron would know more (as he mentioned it here: #18430)

I didn't test whether the dogs attack when you rescue the defector (as I killed them all before). I believe they were invulnerable in the original until triggered, though personally I think it's fine this was removed.

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a-AI.lua:62]: You have several functions with two or three spaces in between them. One space only please.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a.lua:98]: We'll want to check for `player.HasNoRequiredUnits`.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a.lua]: Player and USSR should be capitalized. 

Also, I would prefer `Player `be `Greece `instead. I know there are some old missions that use `Player `instead of the faction name, but all the recent missions have gone with faction names. I plan to get around to changing the holdouts eventually.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a.lua]: Remove space

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a-AI.lua:86]: One space only please.

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a-AI.lua:108]: And here.

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (Smittytron) [inline mods/ra/maps/allies-09a/allies09a-AI.lua:209]: Remove space

Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a-AI.lua:209]: Thank you for your review. Unused space removed. 
[4b4c38e](https://github.com/OpenRA/OpenRA/pull/18695/commits/4b4c38e8bed864fd0e1a1d749bc44fef1960b7b9)

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a-AI.lua:108]: Thank you for your review. Unused space removed. 
[4b4c38e](https://github.com/OpenRA/OpenRA/pull/18695/commits/4b4c38e8bed864fd0e1a1d749bc44fef1960b7b9)

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a-AI.lua:86]: Thank you for your review. Unused space removed. 
[4b4c38e](https://github.com/OpenRA/OpenRA/pull/18695/commits/4b4c38e8bed864fd0e1a1d749bc44fef1960b7b9)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Thank you for your review. Unused space removed. 
[4b4c38e](https://github.com/OpenRA/OpenRA/pull/18695/commits/4b4c38e8bed864fd0e1a1d749bc44fef1960b7b9)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Replace player to Greece, ussr to USSR.
[4b4c38e](https://github.com/OpenRA/OpenRA/pull/18695/commits/4b4c38e8bed864fd0e1a1d749bc44fef1960b7b9)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua:98]: Add "HasNoRequiredUnits"

File: mods/ra/maps/allies-09a/allies09a-AI.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a-AI.lua:62]: Thank you for your review. Unused space removed. 
[4b4c38e](https://github.com/OpenRA/OpenRA/pull/18695/commits/4b4c38e8bed864fd0e1a1d749bc44fef1960b7b9)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: Style nit: Global variables start with an upper case letter: `KosyginType` and `KosyginContacted` (also contact instead of contect.)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: Style nit: Missing space before the `}`.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: Style nit: Please add a line break here.

File: mods/ra/maps/allies-09a/rules.yaml
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/rules.yaml]: This is the default since #18237 and can thus be removed.

File: mods/ra/maps/allies-09a/rules.yaml
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/rules.yaml]: Style nit: No whitespaces before line breaks (`\n`).

Comment (abcdefg30) [reply mods/ra/maps/allies-09a/rules.yaml]: We should also remove some of the new lines to be consistent with the other briefings. I suggest:
> One of Stalin's top atomic strategists, Vladimir Kosygin, wishes to defect. His knowledge of Stalin's atomic strategies is invaluable to us. We will extract him from the Riga compound where he is stationed.\n\nUse a spy to infiltrate the Soviet command center and contact Kosygin. Once he is out of the building, guide him back to your base any way you can.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: This variable appears to be only used here, so can be made `local`.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: `if not KosyginContacted then`

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: `heli` is not used.

File: mods/ra/maps/allies-09a/map.yaml
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/map.yaml]: Style nit: Actor names should start with a capital letter.

File: mods/ra/maps/allies-09a/map.yaml
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/map.yaml]: Same here.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: This needs a line break:
grafik

**Review by abcdefg30 — COMMENTED** [review]
The facing of the map placed actors seems to be wrong (perhaps imported using an older importer version?). We either need to re-import or multiply all facings by 4 (which should be easy using the corresponding update rule with `OpenRA.Utility.exe ra --update-map`).

File: mods/ra/maps/allies-09a/rules.yaml
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/rules.yaml]: The default should be normal.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/rules.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/rules.yaml]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/rules.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/rules.yaml]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/map.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/map.yaml]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/map.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/map.yaml]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/rules.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/rules.yaml]: Fixed at [c9bbea5](https://github.com/OpenRA/OpenRA/commit/c9bbea5b361b63990c5c073f6ac271692cc2ea62)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: This line has trailing whitespaces.

**Review by abcdefg30 — COMMENTED** [review]
Can you squash the commits together into one single commit, please?

File: mods/ra/maps/allies-09a/rules.yaml
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/rules.yaml]: This needs to use `GroundActor` instead of `Ground`.

**Review by abcdefg30 — COMMENTED** [review]
Some other comments:
- Upon contacting Kosygin it is not quite clear where to go with him. We should either specify within the objective to bring him back to the base, or immediately spawn the extraction helicopter (the latter would be better imo; at the moment it is a bit weird to see it fly in and then the mission ends).
- The extraction helicopter should not be selectable.
- When Kosygin leaves the building, he is immediately attacked by Soviet troops, in the original only Dogs attack him. The easiest solution here would be to give him a TargetType only Dogs can attack. (Let me know if you need help with that.)

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (yuantse) [reply mods/ra/maps/allies-09a/allies09a.lua]: Style nit: Removed trailing whitespaces.
squash the commits.

File: mods/ra/maps/allies-09a/rules.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/rules.yaml]: Code fix: Use GroundActor instead of Ground

**Comment by yuantse** [comment]
> Some other comments:
> 
> * Upon contacting Kosygin it is not quite clear where to go with him. We should either specify within the objective to bring him back to the base, or immediately spawn the extraction helicopter (the latter would be better imo; at the moment it is a bit weird to see it fly in and then the mission ends).
> * The extraction helicopter should not be selectable.
> * When Kosygin leaves the building, he is immediately attacked by Soviet troops, in the original only Dogs attack him. The easiest solution here would be to give him a TargetType only Dogs can attack. (Let me know if you need help with that.)

File: mods/ra/maps/allies-09a/rules.yaml
Comment (yuantse) [reply mods/ra/maps/allies-09a/rules.yaml]: Hi @abcdefg30,

I try to set new TargetTypes name for GNRL(Kosygin) in the rules.yaml. After set, Kosygin seems invincible.
Seems dogs need new ValidTargets. But after I set the ValidTargets under DogJaw (also under rules.yaml).
I cannot launch this map.

Maybe I miss something. Could you give me some advice?

GNRL:
	Targetable:
		TargetTypes: Kosygin
	ProximityCaptor:
		Types: Kosygin

DogJaw:
	ValidTargets: Infantry, Kosygin
	Warhead@1Dam: TargetDamage
		Damage: 100000
		ValidTargets: Infantry, Kosygin

Comment (abcdefg30) [reply mods/ra/maps/allies-09a/rules.yaml]: You don't need `ProximityCaptor`. Without seeing the yaml I can't tell much more. You should run the `--check-yaml` command and look at the output. (E.g. run `OpenRA.Utility.exe ra --check-yaml ./mods/ra/maps/allies-09a` from the command line in the OpenRA directory on Windows.)

Comment (abcdefg30) [reply mods/ra/maps/allies-09a/rules.yaml]: Ah wait, `DogJaw` is a weapon, so you can't override it in `rules.yaml`. You need to define a new `weapons.yaml` and do it there. Then add `Weapons: weapons.yaml` to `map.yaml` like here: https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/maps/survival02/map.yaml#L1012

Comment (yuantse) [reply mods/ra/maps/allies-09a/rules.yaml]: Hi @abcdefg30 

After add new weapons.yaml. Only dogs can attack Kosygin. Thanks.

Clarify PrimaryObjective when Kosygin contacted.
Can't select extraction helicopter.
More script to let this map run like the original.

**Comment by Smittytron** [comment]
Played through on hard. Difficulty feels right, good job on this.

Only thing that I saw was that Soviets have a helipad in their base, which gives the player the tech option to build Blackhawks/Chinooks if they capture it. To make things line up with OpenRA's gameplay, we've been removing Soviet helipads from missions. (see conversation in #18122). On this one, the base would look fine if you remove the hpad entirely without replacing it with another building.

I've been working on a handful of missions myself, so if you end up tackling another mission let me know on #4989 or join the OpenRA Discord so we can coordinate efforts.

**Comment by yuantse** [comment]
Hi @Smittytron 

I've removed Helipad from this map.
By the way, I am working on mission scg11ea.

**Review by Smittytron — DISMISSED** [review]
LGTM

**Comment by pchote** [comment]
From Discord:
> abcdefg30 Today at 9:36 AM
> I started a review of 09, but it is broken (mission never ends) and I didn't look into what's going on yet

Adding the changes label so we don't merge this until the problem has been discovered and resolved.

**Comment by yuantse** [comment]
Hi @abcdefg30 

Could you give me more detail about “mission never ends” so that I can duplicate this issue and try to find the problem.

File: mods/ra/maps/allies-09a/allies09a.lua
Comment (abcdefg30) [inline mods/ra/maps/allies-09a/allies09a.lua]: These three should use `AddObjective` instead.

**Review by abcdefg30 — COMMENTED** [review]
I found the problem with the game not ending, it is not related to this PR. We can merge this after the last comment is resolved.

**Review by abcdefg30 — APPROVED** [review]
Pushed a small fixup myself (and rebased).

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/14d24339ac65e77ba450d94771329d399db36280...ca497611670eb64558569a26fd7592b40a047156)

**Pull request merged**
